src/inferencer/inferencer.ts

```diff
@@ -0,0 +1,394 @@
+// inference.ts
+/**
+ * Implements type inference for Source 1 and 2.
+ * @version 0.0.1
+ * @packageDocumentation
+ */
+
+import * as t from './types'
+import { Environment, Scope, Type } from './types'
+import * as es from 'estree'
+import { Context, TypeAnnotatedNode } from '../types'
+import { createEnv, toJson, UNARY_MINUS_OPERATOR } from './util'
+import {
+  ConsequentAlternateMismatchError,
+  InvalidArgumentTypesError,
+  InvalidTestConditionError
+} from '../errors/typeErrors'
+
+function isInternalError(error: any) {
+  return error instanceof t.InternalError
+}
+
+/**
+ * Unifies two types. This might modify the two types.
+ *
+ * @param a The first type to be unified.
+ * @param b The second type to be unified.
+ * @param node
+ */
+
+// tslint:disable-next-line:cyclomatic-complexity
+function unify(a: Type, b: Type) {
+  a = prune(a)
+  b = prune(b)
+
+  // if either is an internal error, we throw immediately
+  if (a instanceof t.InternalError || b instanceof t.InternalError) {
+    throw new t.InternalError()
+  }
+
+  if (a instanceof t.Var) {
+    if (a !== b) {
+      if (occursInType(a, b)) {
+        /**
+         * unify T1 with Pair<T2, T1>
+         */
+        if (b instanceof t.Pair && prune(b.types[1]) === a) {
+          a.instance = new t.List(b.types[0])
+        } else {
+          throw new Error('Recursive unification in types ' + a + ' ' + b)
+        }
+      } else if (a.canContain(b)) {
+        a.instance = b
+      } else if (b instanceof t.Var && b.canContain(a)) {
+        b.instance = a
+      } else {
+        throw new Error('Unable to unify ' + a + ' with ' + b)
+      }
+    }
+  } else if (!(a instanceof t.Var) && b instanceof t.Var) {
+    unify(b, a)
+  } else if (!(a instanceof t.Var || b instanceof t.Var)) {
+    if (a instanceof t.Polymorphic && b instanceof t.Polymorphic) {
```

Comment:
From my understanding, you should not see any polymorphic types in unify I think. You only get polymorphic types in the type environment. When fetching a polymorphic type, you would replace the bound variables with fresh type variables. Within the rules itself you only make use of monomorphic types and only create a polymorphic type when storing back into the type environment.

Reply:
Yeah, it should be called Parameterised instead, in that Functions, Lists, and Pair are all parameterised over multiple types.